### ISSUE TITLE
How to train YOLOv8 with videos

### ISSUE THREAD

**Issue by benniesm** [ultralytics/ultralytics#3357]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am trying to train YOLOv8 classification models on a dataset of many videos. The sequence of the events in the videos are important, therefore breaking them down into individual frames does not seem suitable.
For instance, suppose I have a dataset for many videos, where people bend over to pick up something from the floor (one class), and people doing yoga exercises (another class), I want to be able to make YOLO differentiate between those two human actions, based on the sequence of the movements in a video. And also being able to apply the same ideas to other similar scenarios.
Kindly show how to get this done.
Thanks.

**Comment by glenn-jocher** [comment]
@benniesm hi there,

Thank you for reaching out to us with your question. Training YOLOv8 on video data requires a slightly different approach compared to training on static images. Instead of breaking down the videos into individual frames, you can utilize a technique called video annotation.

To train YOLOv8 with video data, you can use a tool like LabelImg or RectLabel to annotate the videos. You would label each frame of the videos, drawing bounding boxes around the objects of interest (such as people bending or doing yoga exercises) and assigning the corresponding class labels.

When training on the annotated videos, YOLOv8 can learn to detect and differentiate between the different human actions based on the sequence of movements. The temporal information in the video frames allows the model to understand the sequential patterns.

After annotating the videos, you can follow the regular training process for YOLOv8. This involves specifying the dataset path, setting the model configuration, and starting the training process. You can find detailed instructions on how to train YOLOv8 on the Ultralytics YOLOv8 repository documentation.

I hope this helps you get started with training YOLOv8 on your video dataset. If you have any further questions or need additional assistance, please feel free to ask.

Best regards,
Glenn Jocher

**Comment by chijiokeprince92** [comment]
Hi @glenn-jocher,
I wanted to confirm based on your above explanation that the YOLOv8 model will still be trained with frames extracted from the videos but with annotation but not the raw video right?

And if so, how can we make sure that when training, the YOLOv8 model uses the frames sequentially, so that it can also understand sequences that leads to an action?

**Comment by Djkmarco** [comment]
I'm also interested in how to train YOLOv8 for object detection on videos. I've I dataset where the same object appears in several frames, and I think detection could benefit from knowing that it is the same object at different time points instead of just splitting the frames.
@glenn-jocher, is it possible to have a basic colab explaining how to input the data for training and perform such task?

**Comment by glenn-jocher** [comment]
@Djkmarco hi, thank you for your interest in YOLOv8 and object detection on video data.

When training YOLOv8 for video object detection, the common approach is to break down the videos into individual frames and treat each frame as an image during the training process. Therefore, you'd annotate each frame with bounding boxes and labels for the objects you're interested in detecting.

If you want YOLOv8 to recognize that it’s detecting the same objects across different frames, you might benefit from an object tracking system on top of object detection. Object tracking allows the model to understand the continuity of an object across frames, providing it with a notion of temporal consistency.

However, it worth mentioning that training on a video where the same object appears in many frames might introduce a bias in your training set towards this specific object, which might not generalize well to new videos or scenes.

Regarding your request for a basic Colab notebook, although it's a great suggestion, providing such detailed instructions is beyond the scope of this conversation. Nonetheless, you can infer some relevant steps for data preparation from the Ultralytics YOLOv8 documentation.

Overall, training YOLOv8 on video data involves preparing your data correctly (extracting frames, annotating objects), configuring your model, and then starting the training process.

Good luck with your tasks, and let us know if we can offer any further assistance.
  
With best regards,
Glenn Jocher

**Comment by JDGescola** [comment]
Good afternoon. I have a question. I'm currently trying to train YOLOv8 to identify a specific physical exercise being performed, but I've encountered an issue. For example, when detecting a pushup, there's a phase during the exercise where the model might identify that a plank is being performed instead, meaning it can't recognize that the action belongs to a pushup. I would like to know if training the model with videos could solve this problem or if I need to look for another external solution. I am a beginner in this field, so I'm not sure if this would be the best way or if there's another approach.

**Comment by CHinAcSci** [comment]
> @benniesm hi there,
> 
> Thank you for reaching out to us with your question. Training YOLOv8 on video data requires a slightly different approach compared to training on static images. Instead of breaking down the videos into individual frames, you can utilize a technique called video annotation.
> 
> To train YOLOv8 with video data, you can use a tool like LabelImg or RectLabel to annotate the videos. You would label each frame of the videos, drawing bounding boxes around the objects of interest (such as people bending or doing yoga exercises) and assigning the corresponding class labels.
> 
> When training on the annotated videos, YOLOv8 can learn to detect and differentiate between the different human actions based on the sequence of movements. The temporal information in the video frames allows the model to understand the sequential patterns.
> 
> After annotating the videos, you can follow the regular training process for YOLOv8. This involves specifying the dataset path, setting the model configuration, and starting the training process. You can find detailed instructions on how to train YOLOv8 on the Ultralytics YOLOv8 repository documentation.
> 
> I hope this helps you get started with training YOLOv8 on your video dataset. If you have any further questions or need additional assistance, please feel free to ask.
> 
> Best regards, Glenn Jocher

Hello, I have a question about training a model based on yolov8 which can process videos as well as constant frames. The model aims to judge scores for those who like skiing on a ski machine. Based on your ways, we will try to use video annotation but there is a question that how to process our datesets with video annotation? Yolov8 has four kinds of pre-trained model, which should I choose to train my model? classification? I hope you can offer me help. Thank you so much

**Comment by glenn-jocher** [comment]
@CHinAcSci hello,

Thank you for reaching out. For your task, it seems you're looking to perform activity recognition, which involves predicting the action category of a person in a video or sequence of frames. It indeed requires a different approach than static object detection that YOLOv8 performs.

In order to process datasets with video annotation, you have to convert your videos into sequences of frames and annotate each frame using a tool like LabelImg. Each annotation will consist of bounding boxes around the person skiing and labels indicating the action being performed.

However, YOLOv8 is not designed out-of-the-box to use the sequential information of frames in a video for predicting the category of action. YOLOv8 is a state-of-the-art object detection model intended for detecting objects in static images. If you wish to train YOLOv8 on your video data, YOLO treats each frame as a separate, standalone image and does not take into account the sequence of frames.

Since your task involves temporal sequence analysis, you may need to look into action recognition models that are trained to recognize sequences of actions in videos. These models use 3D convolutional networks or recurrent neural networks to track movements across time and are more suited to video classification tasks.

As to your question about which pre-trained model to choose, this depends on your specific use case. The classification model is suited if your task involves categorizing images into one of several predefined classes.

Remember to balance the model size and complexity based on the computational resources you have available. You'll want to choose a model that performs well but also trains in a reasonable time frame given your hardware.

Good luck with your project, and feel free to return if you have more questions.

**Comment by StuffyMonkey** [comment]
What about labels format for object detection on videos? Should it be json file or txt or what? I know that LabelImg or RectLabel and even LabelStudio return label. But it seems like "dirty" json files. For example in LabelStudio in each json there's user information, created_ago and other stuff, while we only neeed 'result' dictionary inside json. Should we extract it manually or the script will preprocess it on his own?

**Comment by glenn-jocher** [comment]
@StuffyMonkey for training YOLOv8 on object detection tasks, the labels for the images (or frames extracted from videos) are typically in a plain text file with a `.txt` format for each image. These text files are concise and contain only the necessary bounding box and class label information.

The `.txt` files for YOLO format generally have one row per object and each row contains:
- Class index (starting from 0)
- x_center and y_center of the bounding box (normalized to the width and height of the image)
- Width and height of the bounding box (also normalized)

The format looks like this:

```
<class_index> <x_center> <y_center> <width> <height>
```

Regarding the annotation tools like LabelImg or LabelStudio, which may produce more complex output formats such as JSON with additional metadata, you would indeed need to preprocess these to extract only the relevant information required for YOLOv8 training. You would typically write a conversion script to parse through the JSON files, extract the 'result' or relevant bounding box and class label data, and then format and save it in the YOLO `.txt` file format for each corresponding frame.

There’s no built-in functionality in YOLOv8 to directly handle raw output from these tools, so preprocessing steps are usually necessary to create a clean dataset for training.

I hope this clarifies your question. If you have any further doubts, don’t hesitate to reach out. Good luck with your training!

**Comment by StuffyMonkey** [comment]
@glenn-jocher Really appreciate your quick response! Final question: as I understand you can't create labels like
`<time_block> <x_center> <y_center> <width> <height>` to train model on videos for object detection (object tracking) task . So you should split video into frames (as you suggested in your response). BUT you can create labels for videos in case of classification task because label will be just txt file with class label. Am I right?

**Comment by glenn-jocher** [comment]
@StuffyMonkey indeed, you are correct in your understanding!

For **object detection** tasks using YOLOv8, you’ll be working with individual frames extracted from videos. Each frame is treated like a still image, and you’ll have corresponding `.txt` label files for each image/frame. These label files don’t include any temporal information (like `<time_block>`) because YOLOv8 doesn’t inherently process temporal sequences. Rather it treats each image independently during both training and inference.

Here’s the breakdown for label files in object detection with YOLOv8:
- Individual frames: Each video frame is saved as an image (e.g., `.jpg` or `.png`).
- Label files: Each frame has an associated `.txt` label file that contains one line for each object in the format `<class_index> <x_center> <y_center> <width> <height>`, with all values normalized relative to the image size.

For **classification** tasks, where you want to label entire videos (each video has one class label), you would indeed be dealing with a different labeling concept. Instead of having a label for each frame, you may have a single `.txt` or even a `.csv` file for the entire dataset that maps video filenames to their corresponding class labels.

In this scenario, the label file for video classification could simply look like:
```
<video_filename1> <class_label>
<video_filename2> <class_label>
```
And so on, without the need for bounding box annotations since you're classifying at the video level rather than detecting objects in individual frames.

It's important to highlight that YOLOv8 is not optimized for video classification tasks in the way that it is for object detection, and if your application involves interpreting sequences (like recognizing a complete physical exercise), models specifically designed for video classification or action recognition should be considered.

Let me know if that brings clarity or if you have any more questions.

**Comment by Choaib-ELMADI** [comment]
Hi, I am trying to build a sign language detection model and I have a set of videos that contains different signs for different meaning. Is it possible to train this model based on these videos without considering each frame on its own because the order of the movements is very important? Thanks

**Comment by glenn-jocher** [comment]
@Choaib-ELMADI hi there! 👋

For a sign language detection task, where the sequence of movements is crucial, splitting videos into individual frames and treating them independently might not capture the temporal dynamics effectively.

YOLOv8 specializes in object detection from static images and treats each image without taking into account the sequence of frames. However, for action or activity recognition like sign language detection, models designed to understand temporal sequences (such as LSTM networks or 3D convolutional networks) might be more suitable.

Although directly training YOLOv8 with videos to recognize sequential actions isn’t feasible, you can still leverage it for robust frame-level feature extraction. Then, you might combine those features in a sequence model to understand the order of movements.

For the initial frame-level sign detection with YOLOv8, you could:
1. Extract frames from videos.
2. Annotate frames with sign language gestures.
3. Train YOLOv8 for gesture detection on these frames.

Then, use a sequence model to interpret the sequence of detected gestures across frames for complete sign language recognition.

I hope this helps! If you have further questions or need more guidance, feel free to ask. Good luck with your project! 🚀

**Comment by pstephenwille** [comment]
great thread.  However, I'm horrified at the idea of annotating each frame.  Is there an easier way to do this other than manually managing each frame one at a time?

**Comment by Choaib-ELMADI** [comment]
@pstephenwille Annotating every frame!!! You must really love pausing HHHHH

**Comment by mzhadigerov** [comment]
The task you are describing is called action recognition or action detection. Yolo is not designed to work with temporal data.

**Comment by pderrenger** [comment]
You're correct. YOLO is primarily designed for object detection in static images and doesn't handle temporal sequences directly. For action recognition, consider using models specifically designed for temporal data, such as LSTM networks, 3D CNNs, or Transformer-based models. These models can capture the sequence of movements in videos, making them more suitable for tasks like sign language detection. If you still want to use YOLO for initial detection, you can extract features from frames and then feed them into a temporal model for action recognition.
